packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -2177,14 +2186,18 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
   late final _WhitespaceDirectionalityFormatter _whitespaceFormatter = _WhitespaceDirectionalityFormatter(textDirection: _textDirection);
 
   void _formatAndSetValue(TextEditingValue value) {
-    // Check if the new value is the same as the current local value, or is the same
-    // as the pre-formatting value of the previous pass (repeat call).
-    final bool textChanged = _value.text != value.text || _value.composing != value.composing;
+    // Only apply input formatters if the text has changed (including uncommited
+    // text in the composing region), or when the user committed the composing
+    // text.
+    // Gboard is very persistent in restoring the composing region. Applying
+    // input formatters on composing region only changes (except clearing the
+    // current composing region) is very infinite-loop-prone: the formatters
+    // will keep trying to modify the composing region while Gboard will keep
+    // trying to restore the original composing region.
+    final bool textChanged = _value.text != value.text
+                          || (!_value.composing.isCollapsed && value.composing.isCollapsed);
```

Comment:
What will happen if a developer does programmatically change the composing region now?  Will the IME still receive the update somehow or will it think that it's using the old composing region?  Or are we saying that developers shouldn't do that (only in TextInputFormatters)?

Reply:
Yeah it's still possible to create an infinite loop in a `TextEditingController` listener callback, but that's like the developer directly fighting the IME, I'm not sure if there's anything we can do to stop that from happening.